enemybar

mjmckee opened the pull request:
enemybar is an add-on that displays the current target in the top center
of the screen for easy viewing.
This is the initial version.

mjmckee committed:
enemybar

enemybar is an add-on that displays the current target in the top center
of the screen for easy viewing.
This is the initial version.

mjmckee committed:
Update enemybar.lua

Added Licensing.

mjmckee committed:
Update gui_settings.lua

Added Licensing.

z16 commented on addons/enemybar/enemybar.lua:
```diff
@@ -0,0 +1,107 @@
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
This part needs to be commented out, or the addon will be unable to load.

z16 commented on addons/enemybar/enemybar.lua:
```diff
@@ -0,0 +1,107 @@
+require('gui_settings')
+
+windower.register_event('load', function()
+	if windower.ffxi.get_info().logged_in then
```
Change the file to indent one level with 4 spaces, instead of 1 tab.

alphaONE2 commented:
Also need to add an entry to addons.xml

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,82 @@
+texts = require('texts')
+
+defaults = {}
+defaults.global = {}
```
You should not add a `global` sub group to settings. This will be generated automatically.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,82 @@
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Same as in the other file, this needs to be commented out.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,82 @@
+defaults.global.font = 'Arial'
+defaults.global.txtSize = 14
+defaults.global.strkSize = 1
+defaults.global.style = 0
```
While this is merely a stylistic suggestion, you should be consistent with your casing. Settings commonly use _PascalCase_. Also, there's no reason to abbreviate _stroke_ to _strk_, this will only confuse some people. `StrokeSize` is a perfectly fine name for this setting.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,82 @@
+	
+	txtMain:stroke_width(settings.global.strkSize)
+	txtMain:stroke_color(50, 50, 50)
+	txtMain:stroke_transparency(127)
```
The `texts` library has an option to automatically generate settings for you, you can look into that. You can check out other addons like _Distance_ to see how they do it. It's really easy to set up.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,82 @@
+	fg_image:size(settings.global.Height, settings.global.Width)
+	fg_image:repeat_xy(1, 1)
+	
+	txtMain:pos(400, 50)
```
Another stylistic choice, but we commonly use _snake_case_ for all our code. But as above, it's up to you if you want to adhere to that, just a suggestion.

mjmckee committed:
Update enemybar.lua

Fixed license to be commented.

mjmckee committed:
Update gui_settings.lua

Commented out license.

mjmckee committed:
Ver. 0.3.1

+Subtarget Support
+Cleaned up code to be more modular and flexible.

mjmckee commented:
I will totally fix all of that and more. I've pushed a new version anyway that fixes some things like the comments on license.

mjmckee committed:
Ver. 0.3.1

+Made some adjustments based on suggestions.

z16 commented on addons/enemybar/targetBar.lua:
```diff
@@ -0,0 +1,143 @@
+					if target.claim_id == 0 then
+						t_text:color(230, 230, 138)
+					else 
+						if target.hpp == 0 then
```
No need to nest these, Lua supports the `elseif` keyword.

mjmckee replied:
This code didn't work correctly anyway. I have just fixed this to not only work correctly but use elseif method to look cleaner.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,72 @@
+defaults.global.tbg_body_path = windower.addon_path.. 'tbg_body.png'
+defaults.global.tfg_body_path = windower.addon_path.. 'tfg_body.png'
+defaults.global.stbg_body_path = windower.addon_path.. 'stbg_body.png'
+defaults.global.stfg_body_path = windower.addon_path.. 'stfg_body.png'
```
I would not define paths in the settings unless you actually want users to change them. This feels more like part of the addon than something configurable, so I would make it static.

mjmckee replied:
I've fixed this on my local version as well as the globals deal. I'm not sure what the actual structure of settings.xml would be so I just guessed based on what one looked like.

alphaONE2 commented on addons/addons.xml:
```diff
@@ -185,6 +185,12 @@
+    <description>This is an addon for Windower4 for FFXI. It creates a big health bar for the target to make it easy to see.</description>
+    <bugtracker>https://github.com/mjmckee/Lua/issues</bugtracker>
+    <support>http://www.ffxiah.com/player/Ragnarok/Pesche</support>
+  </addon>
```
Missing an `<author>` tag here.

mjmckee replied:
Okay, I'll fix this next and then push all my changes so far.

mjmckee committed:
Ver 0.3.2

+Text Color Change is correctly implemented
+Subtarget Color Change added
-messy code

alphaONE2 commented:
It looks like you've fixed pretty much everything here, but your indentation is all over the place. Make sure you've configured your text editor to use 4 spaces instead of tabs. Any good text editor can do this, if you're using Microsoft Notepad, stop and go get [Sublime Text](http://www.sublimetext.com/), [Atom](https://atom.io/), [Notepad++](https://notepad-plus-plus.org/), or literally [anything else](https://en.wikipedia.org/wiki/Comparison_of_text_editors). Then untabify everything; that is convert all the tabs to spaces. Most good editors can do this automatically as well.

mjmckee committed:
Ver. 1.0

+Final and only animation style (based on FF12)
~Changed how the images are used so that the colors can be changed
procedurally
-Removed unneeded stuff
*Final Version

mjmckee commented:
I fixed my editor (Microsoft Code) but I think it the change isn't retroactive. To be honest I'm not sure how to change it retroactively. I don't know what everyone else sees but it looks fine to me. I'll get notepad++ even though I'm not a fan of changing my tools. Anyway, I just pushed my final version (1.0)   before I make a new pull request (I'm not entirely sure if I need to, or how that works) I'll go through it all and make sure it's 4 spaces and not a full tab stop.

_Okay, after further investigation, Code DID in fact replace tab stop to 4 space BUT it was not retroactive so there were some places in old code that stayed tab. Fixed and now will do another pull request_

mjmckee committed:
Ver. 1.0

+Fixed tabs in code so that it's not insanity.

z16 commented on addons/enemybar/targetBar.lua:
```diff
@@ -0,0 +1,102 @@
+tbg_body = images.new(tbg_body_settings)
+tfgg_body = images.new(tfgg_body_settings)
+tfg_body = images.new(tfg_body_settings)
+t_text = texts.new('  ${name|(Name)} - HP: ${hpp|(100)}% ${debug|}', settings.textSettings, settings)
```
Is that `debug` meant to remain in there?

mjmckee replied:
Yes. I figured I'd leave it in for the future.

z16 commented on addons/enemybar/gui_settings.lua:
```diff
@@ -0,0 +1,251 @@
+
+check_claim = function(claim_id)
+    local player = windower.ffxi.get_player()
+    local party = windower.ffxi.get_party()
```
You don't seem to need this variable in your code since you're using `windower.ffxi.get_mob_by_target` instead, so you might wanna remove it. `windower.ffxi.get_party` is a very expensive function call, so you should avoid it where possible. Currently you call it twice on each rendered frame.

mjmckee replied:
Fixed. I'll remember this for the future.

z16 commented on addons/enemybar/subtargetBar.lua:
```diff
@@ -0,0 +1,80 @@
+    if visible == true then
+        local target = windower.ffxi.get_mob_by_target('t')
+        local subtarget = windower.ffxi.get_mob_by_target('st')
+        local player = windower.ffxi.get_player()
```
While `windower.ffxi.get_player` isn't as bad as `windower.ffxi.get_party` it's still rather expensive. You might wanna place it inside the following `if` clause, because it's not needed outside of it.

What some people do, to avoid calling `windower.ffxi.get_player` on each rendered frame is to globalize it and only update it on load and login. This works if you only need things like the player name or ID, since those don't change while logged in, and in your case you only need the ID, so you avoid needlessly calling it many times.

In fact, as a future improvement you may consider caching the target and subtarget IDs as well and switch to using the `target change` event instead of `prerender` to determine those. But that would be a minor improvement and I'd only recommend it if you're out of other things to do, since you'd still need to call them on `prerender` to determine their HP percentage, but you'd only need to call one of them.

mjmckee replied:
I fixed the get_player issue. I moved it inside the if, and cached the id which only really benefits sub-target since I don't use the animation which relies on player.in_combat)

I think as for the second part, I'll just worry about this later.

z16 commented on addons/enemybar/targetBar.lua:
```diff
@@ -0,0 +1,102 @@
+        
+        local target = windower.ffxi.get_mob_by_target('t')
+        local player = windower.ffxi.get_player()
+        local party = windower.ffxi.get_party()
```
Same as above, `windower.ffxi.get_party` does not seem to be needed at all, so you should remove it from here.

mjmckee replied:
Fixed.

z16 commented:
As for the tab thing, you can use a simple text replace mechanism to replace all tabs to spaces, all editors (including VSCode and even Notepad) support. You shouldn't have to switch editors just for this, VSCode is more than enough for this.

But I wanted to write a bit on settings and how you use them. Most things inside the code are fine, and I'd be ready to pull it (the things I mentioned just now were mostly just efficiency improvements, nothing critical), but there are some things in the settings handling that I feel should be improved before we merge this.

Take this line:

``` lua
defaults.pos.x = windower.get_windower_settings().x_res / 2 - defaults.targetBarWidth / 2
```

Default values should be constants and not calculated when first run. These will be saved into the settings, and then reloaded and overwritten on the next load. Now if someone changes resolutions and logs in again, the settings will be off, and they will overwrite the default values, so the bar will be misplaced.

The key to this is to not write things that vary like that to the settings at all. The same goes for settings that depend on other settings, for the same reason, because it can result in a settings mismatch. Like these:

``` lua
stbg_cap_settings.pos.x = settings.pos.x
stbg_cap_settings.pos.y = settings.pos.y
```

In this case someone could have different settings for their characters and when they relog into the other character, the settings for `stbg_cap_settings` will not update anymore.

Only the most basic settings that all other settings depend on should be explicitly specified. Anything that is inferred on load should _only_ be inferred on load and relog of characters. The `config` library has a mechanism for that in its `register` function, and it works like this (example taken from the _Treasury_ addon):

``` lua
s = S{'pass', 'lot', 'drop'}
code = {}
code.pass = S{}
code.lot = S{}
code.drop = S{}

config.register(settings, function(settings_table)
    code.pass = settings_table.Pass:map(table.get+{ids} .. string.lower)
    code.lot = settings_table.Lot:map(table.get+{ids} .. string.lower)
    code.drop = settings_table.Drop:map(table.get+{ids} .. string.lower)
end)
```

The `code` table contains IDs which the addon should pass, lot or drop respectively. The problem is that the addon settings store the items in string form (their english names). But internally it needs the item IDs.

This snippet registers a function with the `config` library. It will be executed every time the settings are reloaded (on the `load`, `login` and `logout` events). This function sets up values based on the settings. In this case, it reads the new settings (from the new character), which is the sole argument to the function. It parses the strings from the settings file and converts them to IDs, which the addon uses internally. That way it's ensured that every time the settings are reloaded the internal state is valid.

You should adapt your code to use the same function. Make one function that initializes all text/prim settings based on the settings that are stored in the file. And settings that should not be stored (like screen resolution dependent settings, because they can vary) can be determined at this point as well instead of being written to the defaults.

Let me know if you're unclear on any of this, I'd be happy to explain in more detail. You can also hit us up on IRC, since it's easier to talk there.

mjmckee commented:
Okay. I'll work on all this tomorrow. Now that it is basically and functionally stable making these little improvements should be a lot easier. I really appreciate the notes they have helped a lot.

_All issues have been addressed and fixed._

mjmckee committed:
Ver. 1.0.1

+Removed excessive calls to player and target
+Correctly setup settings.xml

mjmckee committed:
Ver. 1.0.2

+ Fixes bug with sub target.
+ Fixes bug of persistence with elements on logout.

mjmckee commented:
I fixed a couple bugs and would like to request a new pull, but this one is STILL? open?

z16 commented on addons/enemybar/enemybar.lua:
```diff
@@ -0,0 +1,54 @@
+
+windower.register_event('logout', function(...)
+    -- This is a super cheap fix, but it works. 
+    windower.send_command("input //lua r enemybar");        
```
Fix for what? Also:

``` lua
    windower.send_command('lua r enemybar')
```

mjmckee replied:
Well if you logout while something is targeted (I.E. In the mog house and type /logout to switch characters.) the screen elements will persist. It's "ugly" because I didn't want to spend any more time on this addon than I already have, but still wanted to fix this issue.

z16 commented:
Sorry, I completely forgot about this. Totally my bad, everything looks fine now (except for the thing I posted but that's not really an issue, just a little ugly).
